### PR TITLE
Add opentelemetry-tracing support to materialized

### PR DESCRIPTION
**Pull request by guswynn**
Adds an optional set of environment variables to turn on an unfiltered, experimental setup for a `tracing-opentelemetry` layer that allows us to send data to honeycomb!

### Motivation
  * This PR adds a known-desirable feature.
distributed tracing support for materialized

### Tips for reviewer

This was a journey:
The important parts:

- We are using `native-tls` + `tonic` using `h2`; `h2` requires `alpn` to setup the handshake correctly (https://docs.rs/h2/latest/h2/#handshake)
- I didn't test if the proxy config works, but at least a direct connection worked through the `ProxyConnector` type
- See below for how I tested it
- I decided not to hide this behind a feature, let me know if you want me too, but it make the tracing setup even more messy than it already does
- We are currently sending UNFILTERED data to opentelemetry, so every single trace, including `hyper` internals and what not are sent. we will definitely need to clean this up in the future (or at least sample)

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
Locally I did `MZ_OPENTELEMETRY_ENDPOINT="https://api.honeycomb.io/" MZ_OPENTELEMETRY_HEADERS="x-honeycomb-team=<redacted>,x-honeycomb-dataset=envtest" cargo run --dev

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

I don't think we should reveal this yet to users

### CODE REVIEW

File: src/materialized/src/bin/materialized/main.rs [hunk 1]
```diff
@@ -423,6 +442,76 @@ fn main() {
...
+        // let endpoint = tonic::transport::Endpoint::from_static("https://api.honeycomb.io/");
+        let endpoint = tonic::transport::Endpoint::from_shared(endpoint)?;
+        let channel = endpoint
+            .connect_with_connector_lazy(mz_http_proxy::hyper::connector_with_h2_alpn())
+            .unwrap();
+        let mut mmap = tonic::metadata::MetadataMap::new();
+        /*
+        mmap.append(
+            "x-honeycomb-team",
+            "f00cbe54a4b49537c987dc6a0e833564".parse().unwrap(),
```
Comment (antifuchs) [inline src/materialized/src/bin/materialized/main.rs:476]: Just to make sure this is recognized as a fake key:

```suggestion
            "fffffffffffffffffffffffffffffffff".parse().unwrap(),
```

Comment (guswynn) [reply src/materialized/src/bin/materialized/main.rs]: ugh, so embarrassing

**Comment by benesch** [comment]
Thanks for fighting the fight on this one, @guswynn! Let me PR some pre-work at ya, to clean up some of the stuff you bumped into here.

**Comment by benesch** [comment]
Ok, and just got the other PR out: #11120. That'll allow you to keep all the ALPN goop out of `mz-http-proxy` and in the tracing configuration code. 🎉

**Comment by guswynn** [comment]
@benesch what a rebase!

File: src/materialized/src/bin/materialized/tracing.rs [hunk 2]
```diff
@@ -27,8 +31,85 @@ use mz_ore::metrics::{MetricsRegistry, ThirdPartyMetric};
...
+
+        let stack = stack.with(otel_layer);
+
+        stack.init();
+        Ok(())
+    } else {
+        stack.init();
+        Ok(())
+    }
+}
```
Comment (benesch) [inline src/materialized/src/bin/materialized/tracing.rs:106]: Can I get ya to apply the following style nits? Generally try to `use` things that aren't horribly ambiguous, and keep bounds on generic parameters in a `where` clause unless they're dead simple:

```diff
diff --git a/src/materialized/src/bin/materialized/tracing.rs b/src/materialized/src/bin/materialized/tracing.rs
index 5dca4ff94..581ae5095 100644
--- a/src/materialized/src/bin/materialized/tracing.rs
+++ b/src/materialized/src/bin/materialized/tracing.rs
@@ -19,10 +19,12 @@ use hyper_proxy::ProxyConnector;
 use hyper_tls::HttpsConnector;
 use prometheus::IntCounterVec;
 use tokio::runtime::Runtime;
+use tonic::metadata::{MetadataKey, MetadataMap};
+use tonic::transport::Endpoint;
 use tracing::{Event, Subscriber};
 use tracing_subscriber::filter::{LevelFilter, Targets};
 use tracing_subscriber::fmt;
-use tracing_subscriber::layer::{Context, Layer, SubscriberExt};
+use tracing_subscriber::layer::{Context, Layer, Layered, SubscriberExt};
 use tracing_subscriber::registry::LookupSpan;
 use tracing_subscriber::util::SubscriberInitExt;
 
@@ -48,18 +50,17 @@ fn create_h2_alpn_https_connector() -> ProxyConnector<HttpsConnector<HttpConnect
     )))
 }
 
-fn configure_opentelemetry_and_init<
-    L: tracing_subscriber::layer::Layer<S> + Send + Sync + 'static,
-    S: ::tracing::Subscriber + Send + Sync + 'static,
->(
-    stack: tracing_subscriber::layer::Layered<L, S>,
+fn configure_opentelemetry_and_init<L, S>(
+    stack: Layered<L, S>,
     runtime: &tokio::runtime::Runtime,
     opentelemetry_endpoint: Option<String>,
     opentelemetry_headers: Option<String>,
 ) -> Result<(), anyhow::Error>
 where
-    tracing_subscriber::layer::Layered<L, S>: tracing_subscriber::util::SubscriberInitExt,
-    for<'ls> S: tracing_subscriber::registry::LookupSpan<'ls>,
+    L: Layer<S> + Send + Sync + 'static,
+    S: Subscriber + Send + Sync + 'static,
+    Layered<L, S>: SubscriberInitExt,
+    for<'ls> S: LookupSpan<'ls>,
 {
     if let Some(endpoint) = opentelemetry_endpoint {
         // Setting up opentel in the background requires we are in a tokio-runtime
@@ -67,9 +68,9 @@ where
         let _guard = runtime.enter();
 
         // Manually setup an openssl-backed, h2, proxied `Channel`
-        let endpoint = tonic::transport::Endpoint::from_shared(endpoint)?;
+        let endpoint = Endpoint::from_shared(endpoint)?;
         let channel = endpoint.connect_with_connector_lazy(create_h2_alpn_https_connector())?;
-        let mut mmap = tonic::metadata::MetadataMap::new();
+        let mut mmap = MetadataMap::new();
         // clap guarantees this is set if `opentelemetry_endpoint` is set
         for header in opentelemetry_headers.unwrap().split(',') {
             let mut splits = header.splitn(2, '=');
@@ -80,7 +81,7 @@ where
                 .next()
                 .context("opentelemetry-headers must be of the form key=value")?;
 
-            mmap.insert(tonic::metadata::MetadataKey::from_str(k)?, v.parse()?);
+            mmap.insert(MetadataKey::from_str(k)?, v.parse()?);
         }
 
         let otlp_exporter = opentelemetry_otlp::new_exporter()
```

**Review by benesch — APPROVED** [review]
What a lift! This looks great. Thanks very much for rebasing. Super nice to have all this complexity tucked away in one file.

File: src/coord/src/catalog.rs [hunk 3]
```diff
@@ -201,6 +201,7 @@ impl CatalogState {
             })
     }
 
+    #[tracing::instrument(skip(self))]
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/coord/src/catalog.rs:204]: Its just an example of what kind of instrumentation we can have in the service. We don't have ANY spans, so in honeycomb we will just see random hyper ones unless we provide this as an example, of a reasonably interesting and common callsite

(it shouldn't show up in logs, cause our fmt layer does not log spans right now)

File: src/materialized/src/bin/materialized/main.rs [hunk 4]
```diff
@@ -373,6 +373,24 @@ pub struct Args {
...
 
+    /// The endpoint to send opentelemetry traces to.
+    /// If not provided, tracing is not sent.
+    ///
+    /// You most likely also need to provide
+    ///
+    #[clap(
+        long,
+        env = "MZ_OPENTELEMETRY_ENDPOINT",
+        requires = "opentelemetry-headers",
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/materialized/src/bin/materialized/main.rs:384]: true

originally it was hardcoded as honeycomb, so i was required the dataset and team, but now its up to the user to set it up correctly

File: src/materialized/src/bin/materialized/tracing.rs [hunk 5]
```diff
@@ -27,8 +31,85 @@ use mz_ore::metrics::{MetricsRegistry, ThirdPartyMetric};
...
+    opentelemetry_headers: Option<String>,
+) -> Result<(), anyhow::Error>
+where
+    tracing_subscriber::layer::Layered<L, S>: tracing_subscriber::util::SubscriberInitExt,
+    for<'ls> S: tracing_subscriber::registry::LookupSpan<'ls>,
+{
+    if let Some(endpoint) = opentelemetry_endpoint {
+        // Setting up opentel in the background requires we are in a tokio-runtime
+        // context
+        let _guard = runtime.enter();
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/materialized/src/bin/materialized/tracing.rs:67]: that seems reasonable!

File: src/materialized/src/bin/materialized/tracing.rs (hunk 2 above)
Comment (guswynn) [reply src/materialized/src/bin/materialized/tracing.rs:106]: I differ in opinion here typically, but I can see how this cleans things up here!

File: src/coord/src/catalog.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (guswynn) [reply src/coord/src/catalog.rs:204]: just in case im going to switch this to `trace` level

**Pull request merged**